eyeshade/controllers/accounts/index.js

```diff
@@ -0,0 +1,309 @@
+      type = 'contribution'
+    } else if (type === 'referrals') {
+      type = 'referral'
+    } else {
+      return reply(boom.badData('type must be contributions or referrals'))
+    }
+
+    const query1 = queries.earnings({
+      asc: order === 'asc'
+    })
+
+    const amounts = await runtime.postgres.query(query1, [type, limit])
+    reply(amounts.rows)
+  }
+},
+
+  auth: {
+    strategy: 'simple',
+    mode: 'required'
+  },
+
+  description: 'Used by publishers for retrieving a list of top channel earnings',
+
+  tags: [ 'api', 'publishers' ],
+
+  validate: {
+    params: {
+      type: Joi.string().valid('contributions', 'referrals').required().description('type of earnings')
+    },
+    query: {
+      limit: Joi.number().positive().optional().default(100).description('limit the number of entries returned'),
+      order: orderParam
+    }
+  },
+
+  response: {
+    schema: Joi.array().items(
+       Joi.object().keys({
+         channel: Joi.string(),
+         earnings: Joi.number().description('earnings in BAT'),
+         account_id: Joi.string()
+       })
+     )
+  }
+}
+
+/*
+   GET /v1/accounts/settlements/{type}/total
+*/
+
+v1.getPaidTotals =
+{ handler: (runtime) => {
+  return async (request, reply) => {
+    let { type } = request.params
+    let {
+      order,
+      limit
+    } = request.query
+
+    if (type === 'contributions') {
+      type = 'contribution_settlement'
+    } else if (type === 'referrals') {
+      type = 'referral_settlement'
+    } else {
+      return reply(boom.badData('type must be contributions or referrals'))
+    }
+
+    const query1 = queries.settlements({
+      asc: order === 'asc'
+    })
+
+    const amounts = await runtime.postgres.query(query1, [type, limit])
+    reply(amounts.rows)
+  }
+},
+
+  auth: {
+    strategy: 'simple',
+    mode: 'required'
+  },
+
+  description: 'Used by publishers for retrieving a list of top channels paid out',
+
+  tags: [ 'api', 'publishers' ],
+
+  validate: {
+    params: {
+      type: Joi.string().valid('contributions', 'referrals').required().description('type of payout')
+    },
+    query: {
+      limit: Joi.number().positive().optional().default(100).description('limit the number of entries returned'),
+      order: orderParam
+    }
+  },
+
+  response: {
+    schema: Joi.array().items(
+       Joi.object().keys({
+         channel: Joi.string(),
+         paid: Joi.number().description('amount paid out in BAT'),
+         account_id: Joi.string()
+       })
+     )
+  }
+}
+
+/*
+  PUT /v1/accounts/{payment_id}/transactions/ads/{token_id}
+*/
+
+v1.adTransactions = {
+  handler: (runtime) => async (request, reply) => {
+    const {
+      params,
+      payload
+    } = request
+    const {
+      postgres
+    } = runtime
+    const parameters = {
+      paymentId: params.payment_id,
+      tokenId: params.token_id
+    }
+
+    const found = await transactions.findTransactionByToken(runtime, postgres, parameters)
```

Comment:
I think it's better to try to insert and handle the error case. (Since we have to handle the error case anyway since it's possible for two simultaneous requests to both go past this check)

Reply:
that makes sense. can do